odl/trafos/wavelet.py

```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+            be transformed and on the chosen wavelet basis.
+        """
+        filterlength = int(self.wbasis[9])
+
+        if x.space.ndim == 1:
+            #A copy of x is unavoidable since bwt writes over the given inputs
+            x_cpy = x.copy()
+            x_cpy = x_cpy.asarray()
+            coeff = np.asarray(self.range.element())
+            bwt.wavelet_transform1D(x_cpy.ctypes.data, x_cpy.shape[0],
+                                    filterlength,
+                                    self.nscales,
+                                    coeff.ctypes.data)
+            return coeff
+
+        if x.space.ndim == 2:
+            x_cpy = x.copy()
+            x_cpy = x_cpy.asarray()
+            coeff = np.asarray(self.range.element())
+            bwt.wavelet_transform2D(x_cpy.ctypes.data, x_cpy.shape[0],
+                                    x_cpy.shape[1], filterlength, self.nscales,
+                                    coeff.ctypes.data)
+            return self.range.element(coeff)
+
+        if x.space.ndim == 3:
+            x_cpy = x.copy()
+            x_cpy = x_cpy.asarray()
+            coeff = np.asarray(self.range.element())
+            bwt.wavelet_transform3D(x_cpy.ctypes.data, x_cpy.shape[0],
+                                    x_cpy.shape[1], x_cpy.shape[2],
+                                    filterlength, self.nscales,
+                                    coeff.ctypes.data)
+            return self.range.element(coeff)
+
+    @property
+    def adjoint(self):
+        """The biorthogonal adjoint wavelet transform."""
+        return AdjBiorthWaveletTransform(ran=self.domain,
+                                         nscales=self.nscales,
+                                         wbasis=self.wbasis)
+
+    @property
+    def inverse(self):
+        """The biorthogonal inverse wavelet transform."""
+        return InverseBiorthWaveletTransform(ran=self.domain,
+                                             nscales=self.nscales,
+                                             wbasis=self.wbasis)
+
+    @property
+    def adjointinverse(self):
+        """The inverse of the adjoint of biorthogonal wavelet transform."""
+        return InverseAdjBiorthWaveletTransform(dom=self.domain,
+                                                nscales=self.nscales,
+                                                wbasis=self.wbasis)
+
+
+class AdjBiorthWaveletTransform(Operator):
+    """Discrete adjoint of biorthogonal wavelet transform between L2 space."""
+
+    def __init__(self, ran, nscales, wbasis):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        ran : `DiscreteLpVector` ("image domain")
+
+        nscales : `int`
+            Number of scaling levels in the wavelet transform.
+
+        wbasis : `str`
+            (``'JOSbiorthN'``) J.-O. Stomberg's biorthogonal wavelets
+            where ``N`` refers to the filter length.
```

Comment:
This is again due to making the calling of `WaveletTransform` and `BiorthWaveletTransform` uniform. Also `nscales` refers to the number of scaling levels that can be chosen independently from the choice of wavelets (i.e. almost independently, up to the maximum useful level that is determined from the size of the image and from the chosen wavelet basis). `N` refers to the filter length that are implemented in Jan-Olov's wavelet library and the values are fixed to `1, 3, 5, 7` or `9`. `nscales` is used for both `WaveletTransform` and `BiorthWaveletTransform` but `N` would only be needed for  `BiorthWaveletTransform`